consensus/scheme/rolldpos/proposalpool.go

```diff
@@ -0,0 +1,148 @@
+package rolldpos
+
+import (
+	"sync"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/iotexproject/iotex-core/v2/blockchain/block"
+	"github.com/iotexproject/iotex-core/v2/pkg/log"
+)
+
+// proposalPool is a pool of draft blocks
+type proposalPool struct {
+	// nodes is a map of draft proposal blocks
+	// key is the hash of the block
+	nodes map[hash.Hash256]*block.Block
+	// leaves is a map of tip blocks of forks
+	// key is the hash of the tip block of the fork
+	// value is the timestamp of the block
+	leaves map[hash.Hash256]time.Time
+	// root is the hash of the tip block of the blockchain
+	root hash.Hash256
+	mu   sync.Mutex
+}
+
+func newProposalPool() *proposalPool {
+	return &proposalPool{
+		nodes:  make(map[hash.Hash256]*block.Block),
+		leaves: make(map[hash.Hash256]time.Time),
+	}
+}
+
+func (d *proposalPool) Init(root hash.Hash256) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.root = root
+	log.L().Debug("proposal pool initialized", log.Hex("root", root[:]))
+}
+
+// AddBlock adds a block to the draft pool
+func (d *proposalPool) AddBlock(blk *block.Block) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	// nothing to do if the block already exists
+	hash := blk.HashBlock()
+	if _, ok := d.nodes[hash]; ok {
+		return nil
+	}
+	// it must be a new tip of any fork, or make a new fork
+	prevHash := blk.PrevHash()
+	if _, ok := d.leaves[prevHash]; ok {
+		delete(d.leaves, prevHash)
```

Comment:
if trying to add 2 blocks with same prevHash, then only the first will be added successfully, say
```
b1.PrevHash() = 0x1234, b1.Time() = 17:56:32
b2.PrevHash() = 0x1234, b2.Time() = 17.56.35 (3 seconds later)
```
only b1 will be added (which deletes `d.leaves[0x1234]`, b2 won't be added into the pool
then situation at L126 won't happend? see comments at L126